Individual conditional expectation (ICE)

jdiaz4302 opened the pull request:
Took a somewhat quick stab at ICE - https://christophm.github.io/interpretable-ml-book/ice.html

Local method looking at the effect of changing one feature while holding the others constant. Might help with static importance @SimonTopp. I only applied it to one "batch" input for the RGCN, but it might be worth aggregating more over a whole data set for a more global interpretation. Leads to some cool and intuitive figures.

I was mostly focused on static variables, so it replaces the entire space-time input with 1 value for each grid point estimation, but the results don't look fishy for the dynamic variables either (i.e., rain and air temp). 

Tried to functionalize already and document assumptions

jdiaz4302 committed:
small ICE contribution

SimonTopp commented:
This looks super cool!  I'm off the rest of this week, but I'll take a closer look next Monday/Tuesday.

SimonTopp commented:
Hey @jdiaz4302, just went through the notebook in more detail.  No suggested edits, but a couple high level ideas to think about:
- The ease of interpretation here is great, definitely one of the strong points.  It's interesting to me that this is "local" in time but that we see the influence across all 455 reaches (assuming I'm understanding this right).  I wonder if it'd be helpful to also plot the actual prediction along each reach line too so you can look at it and say "Ok, that's the real world scenario, as we increase or decrease the feature value from that point here is the change we see".  Does that make sense?
- >I only applied it to one "batch" input for the RGCN, but it might be worth aggregating more over a whole data set for a more global interpretation.
- Ya, this is tricky, when interpreting the feature influence I think users need to consider context (i.e. this is the influence of slope during the summer).  I wonder if sampling from different timesteps (or batches depending on if all the batches end at the same time of year) would make things look super noisy or if we'd see consistent patterns in the ICE values.  Might be a way to indirectly get at feature interactions. If the ICE values don't change with things like time of year, then that feature probably isn't interacting with others a ton, but if they do change a lot, we know that there's some amount of feature interaction.
- Re: local vs global interpretations, I see value in both.  Seems like you just need to be explicit regarding if the interpretation is for a target prediction or global influence. Related to the above comment, would a low-hanging fruit approach be to plot the black lines for a random timestep for each reach and then the yellow line could be the mean of all timesteps rather than just the last?  Would probably be a noisier plot and harder to interpret, but maybe an interesting idea.

jdiaz4302 committed:
add several of Simon's ideas

jdiaz4302 commented:
I wanted to address/add most of the things that you said:
* The "local" terminology is awkward here because one input for the RGCN is inherently not local, but it is meant in contrast to entire model/data set behavior.
  * See last bullet point for a small change to make the computation more local.
* Plot actual observations-predictions alongside the lines. Added as small red x's.
* Sample at different time stamps. I added two low effort examples
  * I agree that looking at these curves across different space-time localities would be interesting, but my goal was to write the code, not conduct a case study that individual project scientists can do better.
  * Also agreed about the interactions, but I would like to pursue some more explicit methods for interaction such as H-statistic and 2-variable PDPs (which I've found awkward to approach with our models)
* Plot black lines for random time stamps rather than the final time stamp, then plot yellow line as average of all.
  * I did this at the bottom section of the notebook along with making the local method more local by changing the feature value only for the each segment/black line (while keeping the rest the same). Changing only one segment at a time is very computationally expensive (i.e., `samples * segments` rather than `samples`). 

Notes:
* The actual observations-predictions don't generally fall exactly on the black lines because the whole time series and spatial area of values gets replaced; the last section of the notebook does not replace the whole spatial area, just local - but still whole time series replaced. It's not as simple as "at the same x value, the red point should fall on the black line" because the black line represents an entire time series being scrambled while the red point references that original value at the sampled time stamp (with an unaltered preceding time series). That said, red points and black lines do appear to generally agree. 
* In this prototyping notebook, I left some low-hanging code improvements because I hardcoded 455, 180, and 7 representing `n_segs`, `seq_len`, and `n_vars`. Specifically in the last section which would be somewhat time-consuming to fix and rerun/test.
* In the last section I changed the sampling values from `matplotlib` histogram breaks to deciles (i.e., quantiles by 0.1). This was for practical computation time and mirrors some published work. I'd like to add x-tick indicators for where the sampled values are located but I didn't find a solution that I liked. (e.g., red ticks in figure 4 of https://doi.org/10.1111/gwat.13063)

# Demo figures

## Adding observations

![demo_figure_ice_obs](https://user-images.githubusercontent.com/18230221/179785289-3c02cca5-d0f0-43db-b4b2-15c53434e131.PNG)

## Different time stamps

![demo_figure_ice_obs2](https://user-images.githubusercontent.com/18230221/179785608-28d75ee9-e9be-4638-8cc0-d69e6088e73f.PNG)

![demo_figure_ice_obs3](https://user-images.githubusercontent.com/18230221/179785626-a4b16e6d-1081-40bb-8cd3-a3de9ba49c94.PNG)

## Sample lots of time stamps, plot overall average

![demo_figure_ice_obs4](https://user-images.githubusercontent.com/18230221/179785995-188db5b4-18df-4163-8b68-4df832d886d1.PNG)
